### PR TITLE
Closes #2526: Add arc/hyperbolic Tests to new test framework

### LINKED ISSUE #2526: Add arc/hyperbolic Tests to new test framework
The tests that were added in PR #2447 need to be added to `PROTO_tests/tests/numeric_tests.py`. This file was converted as an example for converting other tests, but we need to make sure we don't lose coverage from the tests added after the conversion.

### PR DESCRIPTION
**Pull request by jaketrookman**
This PR (Closes #2526)
Testing implemented to match #2400 in the new testing framework in `numeric_test.py`.

### CODE REVIEW

File: PROTO_tests/tests/numeric_test.py [hunk 1]
```diff
@@ -197,7 +193,10 @@ def test_cumprod(self, num_type):
...
     @pytest.mark.parametrize("num_type", NO_BOOL)
     def test_sin(self, num_type):
-        na = np.linspace(1, 10, 10).astype(num_type)
+        if num_type == ak.float64:
+            na = np.linspace(0, 10)
+        else:
+            na = np.arange(0, 10).astype(num_type)
         pda = ak.array(na, dtype=num_type)
 
         assert np.allclose(np.sin(na), ak.sin(pda).to_ndarray())
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:202]: I feel like we should update all the `np.allclose` to have `equal_nan=True`. This doesn't seem to factor in for these examples, but would if we add the edge cases at some point

File: PROTO_tests/tests/numeric_test.py [hunk 2]
```diff
@@ -206,13 +205,242 @@ def test_sin(self, num_type):
...
 
+    @pytest.mark.parametrize("num_type", NO_BOOL)
+    def test_tan(self, num_type):
+        if num_type == ak.float64:
+            na = np.linspace(0, 10)
+        else:
+            na = np.arange(0, 10).astype(num_type)
+        pda = ak.array(na, dtype=num_type)
+
+        assert np.allclose(np.tan(na), ak.tan(pda).to_ndarray())
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:226]: I feel like we should update all the `np.allclose` to have `equal_nan=True`. This doesn't seem to factor in for these examples, but would if we add the edge cases at some point

File: PROTO_tests/tests/numeric_test.py [hunk 3]
```diff
@@ -206,13 +205,242 @@ def test_sin(self, num_type):
...
+            pda_num = ak.array(na_num, dtype=num_type)
+        else:
+            na_num = np.arange(10, 0, -1).astype(num_type)
+            pda_num = ak.array(na_num, dtype=num_type)
+        if denom_type == ak.float64:
+            na_denom = np.linspace(0, 10, 10)
+            pda_denom = ak.array(na_denom, dtype=denom_type)
+        else:
+            na_denom = np.arange(0, 10).astype(denom_type)
+            pda_denom = ak.array(na_denom, dtype=denom_type)
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:291]: `pda_num/denom` is the same in all these cases, so we should be able to pull it out.
```python
if num_type == ak.float64:
    na_num = np.linspace(0, 1, 10)
else:
    na_num = np.arange(10, 0, -1).astype(num_type)
if denom_type == ak.float64:
    na_denom = np.linspace(0, 10, 10)
else:
    na_denom = np.arange(0, 10).astype(denom_type)
pda_num = ak.array(na_num, dtype=num_type)
pda_denom = ak.array(na_denom, dtype=denom_type)
```

We could also turn the others into ternaries, but im not sure if that's easier to read or not
```python
na_num = np.linspace(0, 1, 10) if num_type == ak.float64 else np.arange(10, 0, -1).astype(num_type)
na_denom = np.linspace(0, 10, 10) if denom_type == ak.float64 else np.arange(0, 10).astype(denom_type)
pda_num = ak.array(na_num, dtype=num_type)
pda_denom = ak.array(na_denom, dtype=denom_type)
```

File: PROTO_tests/tests/numeric_test.py [hunk 4]
```diff
@@ -206,13 +205,242 @@ def test_sin(self, num_type):
...
+        assert np.allclose(np.arctan2(num[0], na_denom), ak.arctan2(num[0], pda_denom).to_ndarray())
+        assert np.allclose(np.arctan2(na_num, denom[0]), ak.arctan2(pda_num, denom[0]).to_ndarray())
+
+        with pytest.raises(TypeError):
+            ak.arctan2(
+                np.array([range(0, 10)]).astype(num_type), np.array([range(10, 20)]).astype(num_type))
+            ak.arctan2(num[0], np.array([range(10, 20)]).astype(num_type))
+            ak.arctan2(np.array([range(0, 10)]).astype(num_type), denom[0])
+
+        # Edge case: Infinities and Zeros
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:306]: we say `Infinities and Zeros` but i don't think we're testing any zeros

File: PROTO_tests/tests/numeric_test.py [hunk 5]
```diff
@@ -197,7 +193,10 @@ def test_cumprod(self, num_type):
 
     @pytest.mark.parametrize("num_type", NO_BOOL)
     def test_sin(self, num_type):
-        na = np.linspace(1, 10, 10).astype(num_type)
+        if num_type == ak.float64:
+            na = np.linspace(0, 10)
+        else:
+            na = np.arange(0, 10).astype(num_type)
```
[the remark this answers is not in the document]
Comment (Ethan-DeBandi99) [reply PROTO_tests/tests/numeric_test.py:199]: @pierce314159 - I think this is a great idea. 

@jaketrookman - can we make sure to apply this to all tests in the file where applicable?

File: PROTO_tests/tests/numeric_test.py (hunk 1 above)
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:205]: Updated

File: PROTO_tests/tests/numeric_test.py (hunk 2 above)
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:223]: Updated

File: PROTO_tests/tests/numeric_test.py (hunk 3 above)
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py]: Updated

File: PROTO_tests/tests/numeric_test.py (hunk 4 above)
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py]: Updated, the zeros are tested in the main test block

File: PROTO_tests/tests/numeric_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py]: Updated, I was able to get a majority of the trig functions to this format. I did not see very many other places where this structure could be used on the other functions, but definitely looks like a good thing to implement the new testing structure.

**Review by TheOddCzar789 — APPROVED** [review]

File: PROTO_tests/tests/numeric_test.py [hunk 6]
```diff
@@ -10,6 +10,12 @@
 INT_FLOAT = [ak.int64, ak.float64]
 CAST_TYPES = [ak.dtype(t) for t in ak.DTypes]
 
+NP_TRIG_ARRAYS = {
+    ak.int64: np.arange(0, 10, dtype=np.int64),
+    ak.float64: np.linspace(0, 10),
+    ak.uint64: np.arange(0, 10, dtype=np.uint64),
+}
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:17]: this doesn't pick as good of representative examples as I was looking for. We don't get any negative ints or floats, uint's never go into the `> 2**63` range, and we don't test any float edge cases

File: PROTO_tests/tests/numeric_test.py (hunk 1 above)
Comment (pierce314159) [reply PROTO_tests/tests/numeric_test.py:205]: we still need to update this one to use `equal_nan=True`

File: PROTO_tests/tests/numeric_test.py (hunk 2 above)
Comment (pierce314159) [reply PROTO_tests/tests/numeric_test.py:223]: we still need to update this one to use `equal_nan=True`

**Review by pierce314159 — COMMENTED** [review]
Some the issues from my last review are still lingering

File: PROTO_tests/tests/numeric_test.py [hunk 7]
```diff
@@ -197,22 +204,249 @@ def test_cumprod(self, num_type):
...
+        if num_type == ak.float64:
+            na = np.array([np.inf, np.nan, -np.inf])
+            pda = ak.array(na)
+            assert np.allclose(np.arctan(na), ak.arctan(pda).to_ndarray(), equal_nan=True)
+
+    @pytest.mark.parametrize("num_type", NO_BOOL)
+    @pytest.mark.parametrize("denom_type", NO_BOOL)
+    def test_arctan2(self, num_type, denom_type):
+        na_num = NP_TRIG_ARRAYS[num_type]
+        na_denom = np.arange(0, 10).astype(denom_type)
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:291]: I don't love hardcoding the `na_denom` to be fixed to be fixed 0..10 of whatever type. Especially because it means `denom[0]` below ill only ever test the scalar case when it is equal to 0. I think it would be better to take random permutations of the representative arrays, so something the below instead
```python
na_num = np.random.permutation(NP_TRIG_ARRAYS[num_type])
na_denom = np.random.permutation(NP_TRIG_ARRAYS[denom_type])
```

File: PROTO_tests/tests/numeric_test.py [hunk 8]
```diff
@@ -197,22 +204,249 @@ def test_cumprod(self, num_type):
...
+            ak.arctan2(np.array([range(0, 10)]).astype(num_type), denom[0])
+
+        # Edge case: Infinities
+        na_inf = np.array([np.inf, np.nan, -np.inf])
+        pda_inf = ak.array(na_inf)
+        na_fin = np.array([0, 5, 10]).astype(denom_type)
+        pda_fin = ak.array(na_fin)
+        assert np.allclose(
+            np.arctan2(na_inf, na_fin), ak.arctan2(pda_inf, pda_fin).to_ndarray(), equal_nan=True
+        )
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:323]: I think this should be handled by the above once we make the `np.random.permutation` change

File: PROTO_tests/tests/numeric_test.py [hunk 9]
```diff
@@ -197,22 +204,249 @@ def test_cumprod(self, num_type):
...
+            na = np.array([np.inf, np.nan, -np.inf])
+            pda = ak.array(na)
+            assert np.allclose(np.arcsinh(na), ak.arcsinh(pda).to_ndarray(), equal_nan=True)
+
+    @pytest.mark.parametrize("num_type", NO_BOOL)
+    def test_arccosh(self, num_type):
+        if num_type == ak.float64:
+            na = np.linspace(1, 10)
+        else:
+            na = np.arange(1, 10).astype(num_type)
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:399]: is there a reason this test doesn't use `NP_TRIG_ARRAYS`?

File: PROTO_tests/tests/numeric_test.py [hunk 10]
```diff
@@ -10,6 +10,17 @@
...
 
+NP_TRIG_ARRAYS = {
+    ak.int64: np.arange(-5, 5),
+    ak.float64: np.concatenate(
+        [
+            np.linspace(-3.5, 3.5, 5),
+            np.array([np.nan, -np.inf, -0.0, 0.0, np.inf]),
+        ]
+    ),
+    ak.uint64: np.arange(2**64 - 10, 2**64, dtype=np.uint64),
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:21]: No that is a good point. I will add that in there, just in case we want to use it elsewhere in a future test.

File: PROTO_tests/tests/numeric_test.py [hunk 11]
```diff
@@ -197,22 +204,249 @@ def test_cumprod(self, num_type):
...
+
+        assert np.allclose(np.arctan(na), ak.arctan(pda).to_ndarray(), equal_nan=True)
+        with pytest.raises(TypeError):
+            ak.arctan(np.array([range(0, 10)]).astype(num_type))
+
+        # Edge case: infinities and NaNs
+        if num_type == ak.float64:
+            na = np.array([np.inf, np.nan, -np.inf])
+            pda = ak.array(na)
+            assert np.allclose(np.arctan(na), ak.arctan(pda).to_ndarray(), equal_nan=True)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:285]: Updated

File: PROTO_tests/tests/numeric_test.py (hunk 8 above)
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:323]: You are correct that these are now handled, but the next two test are for the vector-scalar and scalar-vector cases. So I will remove the repeated testing.

Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py:323]: Updated

File: PROTO_tests/tests/numeric_test.py [hunk 12]
```diff
@@ -197,22 +205,159 @@ def test_cumprod(self, num_type):
...
+
+        assert np.allclose(
+            np.arctan2(na_num, na_denom), ak.arctan2(pda_num, pda_denom).to_ndarray(), equal_nan=True
+        )
+        assert np.allclose(
+            np.arctan2(num[0], na_denom), ak.arctan2(num[0], pda_denom).to_ndarray(), equal_nan=True
+        )
+        assert np.allclose(
+            np.arctan2(na_num, denom[0]), ak.arctan2(pda_num, denom[0]).to_ndarray(), equal_nan=True
+        )
```
Comment (pierce314159) [inline PROTO_tests/tests/numeric_test.py:280]: how about we update to something like this so we aren't testing the same scalars everytime
```suggestion
        num_scal = pda_num[0]
        denom_scal = pda_denom[0]

        assert np.allclose(
            np.arctan2(na_num, na_denom), ak.arctan2(pda_num, pda_denom).to_ndarray(), equal_nan=True
        )
        assert np.allclose(
            np.arctan2(num_scal, na_denom), ak.arctan2(num_scal, pda_denom).to_ndarray(), equal_nan=True
        )
        assert np.allclose(
            np.arctan2(na_num, denom_scal), ak.arctan2(pda_num, denom_scal).to_ndarray(), equal_nan=True
        )
```

**Review by TheOddCzar789 — APPROVED** [review]

File: PROTO_tests/tests/numeric_test.py [hunk 13]
```diff
@@ -197,22 +205,159 @@ def test_cumprod(self, num_type):
...
+        assert np.allclose(
+            np.arctan2(na_num, denom_scal), ak.arctan2(pda_num, denom_scal).to_ndarray(), equal_nan=True
+        )
+
+        with pytest.raises(TypeError):
+            ak.arctan2(
+                np.array([range(0, 10)]).astype(num_type), np.array([range(10, 20)]).astype(num_type)
+            )
+            ak.arctan2(num[0], np.array([range(10, 20)]).astype(num_type))
+            ak.arctan2(np.array([range(0, 10)]).astype(num_type), denom[0])
```
Comment (Ethan-DeBandi99) [inline PROTO_tests/tests/numeric_test.py:287]: If all of these are expected to raise an error, they need to be checked individually. What is here will raise the expected `TypeError` on the first and the next 2 statements are never executed.

Also, the next 2 statements will not produce a `TypeError` as it stands because `num` and `denom` do not exist. You will raise a `NameError` here.

**Review by Ethan-DeBandi99 — CHANGES_REQUESTED** [review]
The error handling portion of the `arctan2` test is not correct. This needs to be updated before merging. Otherwise, looks good.

**Review by pierce314159 — CHANGES_REQUESTED** [review]

File: PROTO_tests/tests/numeric_test.py [hunk 14]
```diff
@@ -197,22 +205,161 @@ def test_cumprod(self, num_type):
...
+    @pytest.mark.parametrize("denom_type", NO_BOOL)
+    def test_arctan2(self, num_type, denom_type):
+        na_num = np.random.permutation(NP_TRIG_ARRAYS[num_type])
+        na_denom = np.random.permutation(NP_TRIG_ARRAYS[denom_type])
+
+        pda_num = ak.array(na_num, dtype=num_type)
+        pda_denom = ak.array(na_denom, dtype=denom_type)
+
+        num = np.array([1]).astype(num_type)
+        denom = np.array([5]).astype(denom_type)
```
[the remark this answers is not in the document]
Comment (jaketrookman) [reply PROTO_tests/tests/numeric_test.py]: Updated

**Review by Ethan-DeBandi99 — APPROVED** [review]
Looks good

**Review by pierce314159 — APPROVED** [review]
Looks good!
